### ISSUE TITLE
Saving data with add and commit hitting the master database twice

### ISSUE THREAD

**Issue by ksromero** [sqlalchemy/sqlalchemy#5650]
Im trying to change the engine because I'm implementing the redirection of query in read and write databases. I just wonder why it hit the primary database twice.

```
engines = {
    'primary':create_engine'mysql+pymysql://{username}:{password}@{hostname}/{database}',
    'replica':create_engine'mysql+pymysql://{username}:{password}@{hostname}/{database}',
}
```

```
class TestSession(Session):

    def get_bind(self, mapper=None, clause=None):
        if self._flushing:
            logger.debug("Connection is PRIMARY")
            return engines['primary']
        else:
            logger.debug("Connection is REPLICA")
            return engines['replica']
```

This is the query, i just did a simple add and commit for saving the data to database
```
user2 = User(name='user2')
session.add(user2)
session.commit() 
```

and base on the logs
```
[2020-10-16 00:14:32,835] DEBUG in read_and_write: Connection is PRIMARY
[2020-10-16 00:14:32,838] DEBUG in read_and_write: Connection is PRIMARY
[2020-10-16 00:14:32,844] DEBUG in read_and_write: Connection is REPLICA
```

**Complete stack trace, if applicable**

**Versions**
 - OS: Windows 10 Pro
 - Python: 3.8
 - SQLAlchemy: 1.3.20
 - Database: MariaDB
 - DBAPI: PyMysql

**Comment by zzzeek** [comment]
I don't know offhand, so questions like these can be answered using the Python  [pdb](https://docs.python.org/3/library/pdb.html) as well as by setting echo=True on the engine.

We want to see why get_bind() is called twice.  Put a pdb in:

```
class TestSession(Session):

    def get_bind(self, mapper=None, clause=None):
        if self._flushing:
            print("Connection is PRIMARY")
            import pdb
            pdb.set_trace()
            return self.bind
        else:
            print("Connection is REPLICA")
            return self.bind

```

run the program, at each pdb prompt run "where" to see where you are, then "cont" to continue:

```
$ python test3.py 
2020-10-15 21:28:59,013 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2020-10-15 21:28:59,013 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("a")
2020-10-15 21:28:59,013 INFO sqlalchemy.engine.Engine [raw sql] ()
2020-10-15 21:28:59,013 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("a")
2020-10-15 21:28:59,013 INFO sqlalchemy.engine.Engine [raw sql] ()
2020-10-15 21:28:59,014 INFO sqlalchemy.engine.Engine 
CREATE TABLE a (
	id INTEGER NOT NULL, 
	data VARCHAR, 
	PRIMARY KEY (id)
)


2020-10-15 21:28:59,014 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2020-10-15 21:28:59,014 INFO sqlalchemy.engine.Engine COMMIT
Connection is PRIMARY
> /home/classic/dev/sqlalchemy/test3.py(16)get_bind()
-> return self.bind
(Pdb) where
  /home/classic/dev/sqlalchemy/test3.py(34)<module>()
-> s.commit()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(1291)commit()
-> self._transaction.commit(_to_root=self.future)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(709)commit()
-> self._prepare_impl()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(688)_prepare_impl()
-> self.session.flush()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(3105)flush()
-> self._flush(objects)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(3205)_flush()
-> flush_context.execute()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py(442)execute()
-> rec.execute(self)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py(609)execute()
-> util.preloaded.orm_persistence.save_obj(
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(208)save_obj()
-> for (
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(369)_organize_states_for_save()
-> for state, dict_, mapper, connection in _connections_for_states(
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(1707)_connections_for_states()
-> connection = uowtransaction.transaction.connection(base_mapper)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(505)connection()
-> bind = self.session.get_bind(bindkey, **kwargs)
> /home/classic/dev/sqlalchemy/test3.py(16)get_bind()
-> return self.bind
(Pdb) cont
2020-10-15 21:29:00,517 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2020-10-15 21:29:00,520 INFO sqlalchemy.engine.Engine INSERT INTO a (data) VALUES (?)
2020-10-15 21:29:00,520 INFO sqlalchemy.engine.Engine [generated in 0.00051s] (None,)
Connection is PRIMARY
> /home/classic/dev/sqlalchemy/test3.py(16)get_bind()
-> return self.bind
(Pdb) where
  /home/classic/dev/sqlalchemy/test3.py(34)<module>()
-> s.commit()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(1291)commit()
-> self._transaction.commit(_to_root=self.future)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(709)commit()
-> self._prepare_impl()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(688)_prepare_impl()
-> self.session.flush()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(3105)flush()
-> self._flush(objects)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(3205)_flush()
-> flush_context.execute()
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py(442)execute()
-> rec.execute(self)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py(645)execute()
-> util.preloaded.orm_persistence.delete_obj(
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(322)delete_obj()
-> states_to_delete = list(
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(458)_organize_states_for_delete()
-> for state, dict_, mapper, connection in _connections_for_states(
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/persistence.py(1707)_connections_for_states()
-> connection = uowtransaction.transaction.connection(base_mapper)
  /home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/session.py(505)connection()
-> bind = self.session.get_bind(bindkey, **kwargs)
> /home/classic/dev/sqlalchemy/test3.py(16)get_bind()
-> return self.bind
(Pdb) cont
2020-10-15 21:29:02,342 INFO sqlalchemy.engine.Engine COMMIT

```

it looks like it calls upon the bind for both "organize states for save" as well as "organize states for delete".  even though there's nothing to delete here, it calls up the connection that it will use when it builds up this collection as it will associate that connection for each instance. this allows the horizontal sharding session to provide different connections per instance.    The get_bind() method does not necessarily indicate a new connection is being used, as it acts as a key in the current transaction that points to a current connection that came from that engine.

in short,  it's an implementation detail related to the session's features of maintaining multiple engines at once as well as independent connections per object instance.

**Comment by ksromero** [comment]
Thank you. I'm gonna close this one now.
